### ISSUE TITLE
Live queries in CBL iOS not updating on first usage

### ISSUE THREAD

**Issue by jpaas** [couchbase/couchbase-lite-ios#542]
I have an iOS app that uses a few live queries and they all work quite well, except the first time the app is launched. The main view of the app is a table backed by CBLUITableSource that displays each document as a CBLModel. On first launch the list is empty until the user adds the first model. The view has a + button to create new CBLModels. Each time I add a CBLModel, they are persisted, but the livequery backing the list never gets notified of any changes. But this ONLY happens on first launch. If I kill the app and restart it, the models I added before will appear in the list and the live query works from then on and for every app launch after. I can add, delete, and edit and all of those changes will be picked up by the live query. Its only the first time the app is launched that it never works. I've turned on verbose debugging and I don't see any errors. I've attached two logs, one that shows what happens on first launch when I set up the table and create a new CBLModel. The other log shows what happens on the second and subsequent launches when I do the same thing.

I also tried killing the app immediately after initial launch and and restarting before adding my first model. In this case it worked fine.

I found a workaround which is to stop/start the livequery after the query gets updated in the 
couchTableSource:updateFromQuery:previousRows: method. This seems to to allow it receive all future notifications.

This is a swift project running on iOS8 and using a manually built head version of couchbase-lite-ios. I had to build it manually because I needed 1.0.3.1 before it was available on cocoapods. It sure would be nice if the podspec was part of the repo, so that it can be used like most other cocoapods...

```pod 'couchbase-lite-ios', git: 'https://github.com/couchbase/couchbase-lite-ios.git```

Here's the log from first launch:
```
2014-11-28 12:44:17.847 [INFO ] tapshotpro[5639:607] AppDelegate:140 - App started
2014-11-28 12:44:17.892 [INFO ] tapshotpro[5639:607] MigrationManager:29 - Last migration: 0
2014-11-28 12:44:17.892 [INFO ] tapshotpro[5639:607] MigrationManager:32 - Performing migration: 1
2014-11-28 12:44:17.909 [INFO ] tapshotpro[5639:607] TeamsViewController:35 - viewDidLoad
12:44:17.914| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: Async query tapshotpro/kCblViewTeamsByName...
12:44:17.915| CBLDatabase: Created CBLManager[0x7c8cfb50 /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/6B2DF02F-EF73-49F5-93B5-F2BF447AD6AD/Library/Application Support/CouchbaseLite]
12:44:17.915| CBLDatabase: CBLManager[0x7c8a9eb0 /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/6B2DF02F-EF73-49F5-93B5-F2BF447AD6AD/Library/Application Support/CouchbaseLite] created CBL_Server[0x7c8d0210] (with CBLManager[0x7c8cfb50 /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/6B2DF02F-EF73-49F5-93B5-F2BF447AD6AD/Library/Application Support/CouchbaseLite])
12:44:18.111‖ CBLDatabase: Opening CBLDatabase[<0x7ae99250>tapshotpro]
12:44:18.111‖ CBLDatabase: Open /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/6B2DF02F-EF73-49F5-93B5-F2BF447AD6AD/Library/Application Support/CouchbaseLite/tapshotpro.cblite (flags=200006)
12:44:18.112‖ CBLDatabase: Begin transaction (level 1)...
12:44:18.112‖ ViewVerbose:     kCblViewTeamsByName last indexed at #0
12:44:18.113‖ ViewVerbose: #2: map "FB334942-4BA7-4823-A9F4-2B7E846425FE" for view kCblViewTeamsByName...
12:44:18.113‖ CBLDatabase: Commit transaction (level 1)
12:44:18.129| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: Will update after 0.285453 sec...
12:44:18.129| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: ...async query finished (0 rows)
12:44:18.129| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: ...Rows changed! (now 0)
2014-11-28 12:44:18.130 [INFO ] tapshotpro[5639:607] TeamsViewController:72 - teams query changed
2014-11-28 12:44:18.130 [INFO ] tapshotpro[5639:607] TeamsViewController:123 - teams live query updated
2014-11-28 12:44:18.133 [INFO ] tapshotpro[5639:607] TeamsViewController:89 - No team selected
2014-11-28 12:44:25.170 [INFO ] tapshotpro[5639:607] TeamViewController:41 - viewDidLoad
2014-11-28 12:44:34.180 [INFO ] tapshotpro[5639:607] TeamViewController:64 - savePressed
12:44:34.180| CBLModel: New team init
12:44:34.180| CBLModel: New team made new document
12:44:34.180| CBLModel: New team .type := "Team"
12:44:34.181| CBLModel: New team .name := "Warriors"
12:44:34.181| CBLModel: New team Saving <- {
    "_id" = "0EAF904F-CBAF-4BF1-836A-16B52A60D56D";
    name = Warriors;
    type = Team;
}
12:44:34.181| CBLDatabase: PUT rev={0EAF904F-CBAF-4BF1-836A-16B52A60D56D #(null)}, prevRevID=(null), allowConflict=0
12:44:34.181| CBLDatabase: Begin transaction (level 1)...
12:44:34.181| CBLDatabase: Commit transaction (level 1)
12:44:34.182| CBLDatabase: Added: {0EAF904F-CBAF-4BF1-836A-16B52A60D56D #1-aa9c2b95205510cacf5873b401ab4a55} (seq=3)
12:44:34.182| CBLDatabase: CBLDatabase[<0x7c8aae80>tapshotpro]: Posting change notifications: seq 3
12:44:34.182| CBLDatabase: CBLDatabase[<0x7ae99250>tapshotpro]: Notified of 1 changes by CBLDatabase[<0x7c8aae80>tapshotpro]
12:44:34.182‖ CBLDatabase: CBLDatabase[<0x7ae99250>tapshotpro]: Posting change notifications: seq (3)
12:44:34.182| CBLModel: [name: Warriors, type: Team, _id: 0EAF904F-CBAF-4BF1-836A-16B52A60D56D, _rev: 1-aa9c2b95205510cacf5873b401ab4a55] Saved as rev 1-aa9c2b95205510cacf5873b401ab4a55
2014-11-28 12:44:34.184 [INFO ] tapshotpro[5639:607] TeamViewController:76 - Saved: [name: Warriors, type: Team, _id: 0EAF904F-CBAF-4BF1-836A-16B52A60D56D, _rev: 1-aa9c2b95205510cacf5873b401ab4a55]
12:44:34.539| CBLModel: [name: Warriors, type: Team, _id: 0EAF904F-CBAF-4BF1-836A-16B52A60D56D, _rev: 1-aa9c2b95205510cacf5873b401ab4a55] dealloc
```

And on subsequent launches:
```
2014-11-28 12:40:49.593 [INFO ] tapshotpro[5537:607] AppDelegate:140 - App started
2014-11-28 12:40:49.610 [INFO ] tapshotpro[5537:607] MigrationManager:29 - Last migration: 1
2014-11-28 12:40:49.635 [INFO ] tapshotpro[5537:607] TeamsViewController:35 - viewDidLoad
12:40:49.641| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: Async query tapshotpro/kCblViewTeamsByName...
12:40:49.641| CBLDatabase: Created CBLManager[0x78e964f0 /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/93172490-A955-4EB2-89BF-E37F59D850BA/Library/Application Support/CouchbaseLite]
12:40:49.641| CBLDatabase: CBLManager[0x79cb3600 /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/93172490-A955-4EB2-89BF-E37F59D850BA/Library/Application Support/CouchbaseLite] created CBL_Server[0x78e96bb0] (with CBLManager[0x78e964f0 /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/93172490-A955-4EB2-89BF-E37F59D850BA/Library/Application Support/CouchbaseLite])
12:40:49.653‖ CBLDatabase: Opening CBLDatabase[<0x79b3fbb0>tapshotpro]
12:40:49.653‖ CBLDatabase: Open /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/93172490-A955-4EB2-89BF-E37F59D850BA/Library/Application Support/CouchbaseLite/tapshotpro.cblite (flags=200006)
12:40:49.654‖ CBLDatabase: Begin transaction (level 1)...
12:40:49.654‖ CBLDatabase: Commit transaction (level 1)
12:40:49.672| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: ...async query finished (0 rows)
12:40:49.672| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: ...Rows changed! (now 0)
2014-11-28 12:40:49.672 [INFO ] tapshotpro[5537:607] TeamsViewController:72 - teams query changed
2014-11-28 12:40:49.672 [INFO ] tapshotpro[5537:607] TeamsViewController:123 - teams live query updated
2014-11-28 12:40:49.675 [INFO ] tapshotpro[5537:607] TeamsViewController:89 - No team selected
2014-11-28 12:40:51.123 [INFO ] tapshotpro[5537:607] TeamViewController:41 - viewDidLoad
2014-11-28 12:40:54.475 [INFO ] tapshotpro[5537:607] TeamViewController:64 - savePressed
12:40:54.475| CBLModel: New team init
12:40:54.475| CBLModel: New team made new document
12:40:54.475| CBLModel: New team .type := "Team"
12:40:54.475| CBLModel: New team .name := "warriors"
12:40:54.476| CBLModel: New team Saving <- {
    "_id" = "E2651779-91E6-4700-9357-2209AF21E7ED";
    name = warriors;
    type = Team;
}
12:40:54.476| CBLDatabase: PUT rev={E2651779-91E6-4700-9357-2209AF21E7ED #(null)}, prevRevID=(null), allowConflict=0
12:40:54.476| CBLDatabase: Begin transaction (level 1)...
12:40:54.476| CBLDatabase: Commit transaction (level 1)
12:40:54.477| CBLDatabase: Added: {E2651779-91E6-4700-9357-2209AF21E7ED #1-4e233bdf49d013ecf0308fbb1fb731fd} (seq=3)
12:40:54.477| CBLDatabase: CBLDatabase[<0x79cb44b0>tapshotpro]: Posting change notifications: seq 3
12:40:54.477| CBLDatabase: CBLDatabase[<0x79b3fbb0>tapshotpro]: Notified of 1 changes by CBLDatabase[<0x79cb44b0>tapshotpro]
12:40:54.477‖ CBLDatabase: CBLDatabase[<0x79b3fbb0>tapshotpro]: Posting change notifications: seq (3)
12:40:54.477| CBLModel: [name: warriors, type: Team, _id: E2651779-91E6-4700-9357-2209AF21E7ED, _rev: 1-4e233bdf49d013ecf0308fbb1fb731fd] Saved as rev 1-4e233bdf49d013ecf0308fbb1fb731fd
2014-11-28 12:40:54.479 [INFO ] tapshotpro[5537:607] TeamViewController:76 - Saved: [name: warriors, type: Team, _id: E2651779-91E6-4700-9357-2209AF21E7ED, _rev: 1-4e233bdf49d013ecf0308fbb1fb731fd]
12:40:54.480| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: Will update after 0 sec...
12:40:54.480| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: Async query tapshotpro/kCblViewTeamsByName...
12:40:54.480‖ CBLDatabase: Begin transaction (level 1)...
12:40:54.480‖ ViewVerbose:     kCblViewTeamsByName last indexed at #2
12:40:54.481‖ ViewVerbose: #3: map "E2651779-91E6-4700-9357-2209AF21E7ED" for view kCblViewTeamsByName...
12:40:54.481‖ ViewVerbose:     emit("warriors", {"name":"warriors","type":"Team","_id":"E2651779-91E6-4700-9357-2209AF21E7ED","_rev":"1-4e233bdf49d013ecf0308fbb1fb731fd","_local_seq":3})
12:40:54.481‖ CBLDatabase: Commit transaction (level 1)
12:40:54.482‖ ViewVerbose: Query kCblViewTeamsByName: Found row with key="warriors", value={"name":"warriors","type":"Team","_id":"E2651779-91E6-4700-9357-2209AF21E7ED","_rev":"1-4e233bdf49d013ecf0308fbb1fb731fd","_local_seq":3}, id="E2651779-91E6-4700-9357-2209AF21E7ED"
12:40:54.482| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: ...async query finished (1 rows)
12:40:54.482| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: ...Rows changed! (now 1)
2014-11-28 12:40:54.482 [INFO ] tapshotpro[5537:607] TeamsViewController:72 - teams query changed
```

**Comment by pasin** [comment]
Closing as no response. Feel free to reopen the issue if you have updates or more info regarding the issue.

**Comment by jpaas** [comment]
Sorry I haven't had time to try and create an isolated app that demonstrates this issue. Instead I developed a workaround which was to stop and restart the query on first app launch. I can reproduce it easily with my app in both the simulator and on a device. I captured the logs with verbose logging on to compare the first launch vs the second. One thing I realized is that on first launch, the view has to be indexed before it is used by the query. I wonder if that is causing some sort of async issue? Line 12 of first_launch.txt has this extra line when the query is being set up:

```
18:14:55.537| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: Will update after 0.434258 sec...
```

first_launch.txt:
```
2015-02-16 18:14:55.459 [INFO ] tapshotpro[61423:607] TeamsViewController:37 - viewDidLoad
18:14:55.471| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: Async query tapshotpro/kCblViewTeamsByName...
18:14:55.471| CBLDatabase: Created CBLManager[0x78eae810 /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/039857E9-0053-4CF1-9C78-4379355ECE38/Library/Application Support/CouchbaseLite]
18:14:55.472| CBLDatabase: CBLManager[0x7996c2d0 /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/039857E9-0053-4CF1-9C78-4379355ECE38/Library/Application Support/CouchbaseLite] created CBL_Server[0x78eadf10] (with CBLManager[0x78eae810 /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/039857E9-0053-4CF1-9C78-4379355ECE38/Library/Application Support/CouchbaseLite])
18:14:55.472‖ CBLDatabase: Opening CBLDatabase[<0x78eaeee0>tapshotpro]
18:14:55.472‖ CBLDatabase: Open /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/039857E9-0053-4CF1-9C78-4379355ECE38/Library/Application Support/CouchbaseLite/tapshotpro.cblite (flags=200006)
18:14:55.501‖ CBLDatabase: Begin transaction (level 1)...
18:14:55.502‖ ViewVerbose:     kCblViewTeamsByName last indexed at #0
18:14:55.502‖ ViewVerbose: #4: map "F5DACAE9-0670-4877-B118-22A0EE414DB7" for view kCblViewTeamsByName...
18:14:55.502‖ CBLDatabase: Commit transaction (level 1)
2015-02-16 18:14:55.537 [ERROR] tapshotpro[61423:607] AppDelegate:88 - didFailToRegisterForRemoteNotificationsWithError: Error Domain=NSCocoaErrorDomain Code=3010 "remote notifications are not supported in the simulator" UserInfo=0x78e882d0 {NSLocalizedDescription=remote notifications are not supported in the simulator}
18:14:55.537| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: Will update after 0.434258 sec...
18:14:55.537| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: ...async query finished (0 rows)
18:14:55.537| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: ...Rows changed! (now 0)
2015-02-16 18:14:55.538 [INFO ] tapshotpro[61423:607] TeamsViewController:121 - Live query updated with 0 rows
2015-02-16 18:14:55.540 [INFO ] tapshotpro[61423:607] TeamsViewController:87 - No team selected
2015-02-16 18:14:55.543 [INFO ] tapshotpro[61423:607] WelcomeViewController:38 - viewDidLoad
2015-02-16 18:15:00.355 [INFO ] tapshotpro[61423:607] TeamViewController:41 - viewDidLoad
2015-02-16 18:15:03.548 [INFO ] tapshotpro[61423:607] TeamViewController:64 - savePressed
18:15:03.550| CBLModel: New team init
18:15:03.550| CBLModel: New team made new document
18:15:03.550| CBLModel: New team .gameTypeFilter := "31"
18:15:03.550| CBLModel: New team .showInactivePlayers := "0"
18:15:03.550| CBLModel: New team .type := "Team"
18:15:03.550| CBLModel: New team .name := "Leafs"
18:15:03.551| CBLModel: New team Saving <- {
    "_id" = "87CEECEB-C280-4389-80BD-CB65C270829D";
    gameTypeFilter = 31;
    name = Leafs;
    showInactivePlayers = 0;
    type = Team;
}
18:15:03.551| CBLDatabase: PUT rev={87CEECEB-C280-4389-80BD-CB65C270829D #(null)}, prevRevID=(null), allowConflict=0
18:15:03.551| CBLDatabase: Begin transaction (level 1)...
18:15:03.551| CBLDatabase: Commit transaction (level 1)
18:15:03.552| CBLDatabase: Added: {87CEECEB-C280-4389-80BD-CB65C270829D #1-649e533cfa9c4aa19a5fd996b2fd2674} (seq=5)
18:15:03.552| CBLDatabase: CBLDatabase[<0x79c6c210>tapshotpro]: Posting change notifications: seq 5
18:15:03.552| CBLDatabase: CBLDatabase[<0x78e8dc10>tapshotpro]: Notified of 1 changes by CBLDatabase[<0x79c6c210>tapshotpro]
18:15:03.552| CBLDatabase: CBLDatabase[<0x78eaeee0>tapshotpro]: Notified of 1 changes by CBLDatabase[<0x79c6c210>tapshotpro]
18:15:03.552‖ CBLDatabase: CBLDatabase[<0x78e8dc10>tapshotpro]: Posting change notifications: seq (5)
18:15:03.552‖ CBLDatabase: CBLDatabase[<0x78eaeee0>tapshotpro]: Posting change notifications: seq (5)
18:15:03.552| CBLModel: [name: Leafs, gameTypeFilter: 31, showInactivePlayers: 0, type: Team, _id: 87CEECEB-C280-4389-80BD-CB65C270829D, _rev: 1-649e533cfa9c4aa19a5fd996b2fd2674] Saved as rev 1-649e533cfa9c4aa19a5fd996b2fd2674
2015-02-16 18:15:03.554 [INFO ] tapshotpro[61423:607] TeamViewController:78 - Saved: [name: Leafs, gameTypeFilter: 31, showInactivePlayers: 0, type: Team, _id: 87CEECEB-C280-4389-80BD-CB65C270829D, _rev: 1-649e533cfa9c4aa19a5fd996b2fd2674]
18:15:03.554| CBLDatabase: Begin transaction (level 1)...
18:15:03.554| ViewVerbose:     kCblViewTeamsByName last indexed at #4
18:15:03.555| ViewVerbose: #5: map "87CEECEB-C280-4389-80BD-CB65C270829D" for view kCblViewTeamsByName...
18:15:03.555| ViewVerbose:     emit("Leafs", {"_id":"87CEECEB-C280-4389-80BD-CB65C270829D","type":"Team","_local_seq":5,"name":"Leafs","_rev":"1-649e533cfa9c4aa19a5fd996b2fd2674","gameTypeFilter":31,"showInactivePlayers":false})
18:15:03.555| CBLDatabase: Commit transaction (level 1)
18:15:03.556| ViewVerbose: Query kCblViewTeamsByName: Will reduce row with key="Leafs", value={"_id":"87CEECEB-C280-4389-80BD-CB65C270829D","type":"Team","_local_seq":5,"name":"Leafs","_rev":"1-649e533cfa9c4aa19a5fd996b2fd2674","gameTypeFilter":31,"showInactivePlayers":false}
18:15:03.556| ViewVerbose: Query kCblViewTeamsByName: Reduced to key=null, value=1
18:15:03.911| CBLModel: [name: Leafs, gameTypeFilter: 31, showInactivePlayers: 0, type: Team, _id: 87CEECEB-C280-4389-80BD-CB65C270829D, _rev: 1-649e533cfa9c4aa19a5fd996b2fd2674] dealloc
```

second_launch.txt:
```
2015-02-16 18:16:22.127 [INFO ] tapshotpro[61493:607] TeamsViewController:37 - viewDidLoad
18:16:22.138| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: Async query tapshotpro/kCblViewTeamsByName...
18:16:22.138| CBLDatabase: Created CBLManager[0x7aeaf8c0 /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/438E1246-D549-439E-8E2A-0A8680B46E28/Library/Application Support/CouchbaseLite]
18:16:22.138| CBLDatabase: CBLManager[0x7c3b3990 /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/438E1246-D549-439E-8E2A-0A8680B46E28/Library/Application Support/CouchbaseLite] created CBL_Server[0x7aeafba0] (with CBLManager[0x7aeaf8c0 /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/438E1246-D549-439E-8E2A-0A8680B46E28/Library/Application Support/CouchbaseLite])
18:16:22.139‖ CBLDatabase: Opening CBLDatabase[<0x7c482560>tapshotpro]
18:16:22.139‖ CBLDatabase: Open /Users/jpaas/Library/Developer/CoreSimulator/Devices/58A0CA81-23B3-4B2D-B4BA-CF34D29DD597/data/Containers/Data/Application/438E1246-D549-439E-8E2A-0A8680B46E28/Library/Application Support/CouchbaseLite/tapshotpro.cblite (flags=200006)
18:16:22.156‖ CBLDatabase: Begin transaction (level 1)...
18:16:22.156‖ CBLDatabase: Commit transaction (level 1)
18:16:22.156‖ ViewVerbose: Query kCblViewTeamsByName: Found row with key="Leafs", value={"_id":"87CEECEB-C280-4389-80BD-CB65C270829D","type":"Team","_local_seq":5,"name":"Leafs","_rev":"1-649e533cfa9c4aa19a5fd996b2fd2674","gameTypeFilter":31,"showInactivePlayers":false}, id="87CEECEB-C280-4389-80BD-CB65C270829D"
2015-02-16 18:16:22.215 [ERROR] tapshotpro[61493:607] AppDelegate:88 - didFailToRegisterForRemoteNotificationsWithError: Error Domain=NSCocoaErrorDomain Code=3010 "remote notifications are not supported in the simulator" UserInfo=0x7ae97aa0 {NSLocalizedDescription=remote notifications are not supported in the simulator}
18:16:22.215| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: ...async query finished (1 rows)
18:16:22.215| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: ...Rows changed! (now 1)
2015-02-16 18:16:22.215 [INFO ] tapshotpro[61493:607] TeamsViewController:121 - Live query updated with 1 rows
2015-02-16 18:16:22.217 [INFO ] tapshotpro[61493:607] TeamsViewController:239 - Selecting team 87CEECEB-C280-4389-80BD-CB65C270829D at row 0 
18:16:22.218| CBLModel: tapshotpro.Team initWithDocument: CBLDocument[87CE..829D] @0x7c4812e0
2015-02-16 18:16:22.220 [INFO ] tapshotpro[61493:607] TeamsViewController:170 - Selected team: [name: Leafs, gameTypeFilter: 31, showInactivePlayers: 0, type: Team, _id: 87CEECEB-C280-4389-80BD-CB65C270829D, _rev: 1-649e533cfa9c4aa19a5fd996b2fd2674]
2015-02-16 18:16:22.228 [INFO ] tapshotpro[61493:607] TeamsViewController:90 - Selected team for segue: [name: Leafs, gameTypeFilter: 31, showInactivePlayers: 0, type: Team, _id: 87CEECEB-C280-4389-80BD-CB65C270829D, _rev: 1-649e533cfa9c4aa19a5fd996b2fd2674]
18:16:22.228| CBLDatabase: Begin transaction (level 1)...
18:16:22.228| CBLDatabase: Commit transaction (level 1)
2015-02-16 18:16:22.228 [INFO ] tapshotpro[61493:607] TeamTabBarController:64 - Setting games tab enabled to : false
2015-02-16 18:16:22.238 [INFO ] tapshotpro[61493:607] PlayersViewController:34 - viewDidLoad
2015-02-16 18:16:22.296 [INFO ] tapshotpro[61493:607] PlayersViewController:49 - viewDidAppear
18:16:22.296| Query: CBLLiveQuery[kCblViewActivePlayersByNumber, start=["87CEECEB-C280-4389-80BD-CB65C270829D"], end=["87CEECEB-C280-4389-80BD-CB65C270829D","￿","￿"], mapOnly=YES]: Async query tapshotpro/kCblViewActivePlayersByNumber...
18:16:22.296‖ CBLDatabase: Begin transaction (level 1)...
18:16:22.296‖ CBLDatabase: Commit transaction (level 1)
18:16:22.299| Query: CBLLiveQuery[kCblViewActivePlayersByNumber, start=["87CEECEB-C280-4389-80BD-CB65C270829D"], end=["87CEECEB-C280-4389-80BD-CB65C270829D","￿","￿"], mapOnly=YES]: ...async query finished (0 rows)
18:16:22.299| Query: CBLLiveQuery[kCblViewActivePlayersByNumber, start=["87CEECEB-C280-4389-80BD-CB65C270829D"], end=["87CEECEB-C280-4389-80BD-CB65C270829D","￿","￿"], mapOnly=YES]: ...Rows changed! (now 0)
2015-02-16 18:16:22.299 [INFO ] tapshotpro[61493:607] PlayersViewController:158 - Live query updated with 0 rows
2015-02-16 18:16:22.300 [INFO ] tapshotpro[61493:607] WelcomeViewController:38 - viewDidLoad
2015-02-16 18:16:35.462 [INFO ] tapshotpro[61493:607] TeamsViewController:64 - addPressed
2015-02-16 18:16:35.467 [INFO ] tapshotpro[61493:607] TeamViewController:41 - viewDidLoad
2015-02-16 18:16:37.920 [INFO ] tapshotpro[61493:607] TeamViewController:64 - savePressed
18:16:37.920| CBLModel: New team init
18:16:37.920| CBLModel: New team made new document
18:16:37.920| CBLModel: New team .gameTypeFilter := "31"
18:16:37.921| CBLModel: New team .showInactivePlayers := "0"
18:16:37.921| CBLModel: New team .type := "Team"
18:16:37.921| CBLModel: New team .name := "Habs"
18:16:37.921| CBLModel: New team Saving <- {
    "_id" = "14B53188-5AC0-451D-9104-4663108C3BDB";
    gameTypeFilter = 31;
    name = Habs;
    showInactivePlayers = 0;
    type = Team;
}
18:16:37.921| CBLDatabase: PUT rev={14B53188-5AC0-451D-9104-4663108C3BDB #(null)}, prevRevID=(null), allowConflict=0
18:16:37.921| CBLDatabase: Begin transaction (level 1)...
18:16:37.922| CBLDatabase: Commit transaction (level 1)
18:16:37.923| CBLDatabase: Added: {14B53188-5AC0-451D-9104-4663108C3BDB #1-643db5608d6cce93ad21003fe9e6e886} (seq=6)
18:16:37.923| CBLDatabase: CBLDatabase[<0x7c39dea0>tapshotpro]: Posting change notifications: seq 6
18:16:37.923| CBLDatabase: CBLDatabase[<0x7c782040>tapshotpro]: Notified of 1 changes by CBLDatabase[<0x7c39dea0>tapshotpro]
18:16:37.923| CBLDatabase: CBLDatabase[<0x7c482560>tapshotpro]: Notified of 1 changes by CBLDatabase[<0x7c39dea0>tapshotpro]
18:16:37.923‖ CBLDatabase: CBLDatabase[<0x7c782040>tapshotpro]: Posting change notifications: seq (6)
18:16:37.923‖ CBLDatabase: CBLDatabase[<0x7c482560>tapshotpro]: Posting change notifications: seq (6)
18:16:37.923| CBLModel: [name: Habs, gameTypeFilter: 31, showInactivePlayers: 0, type: Team, _id: 14B53188-5AC0-451D-9104-4663108C3BDB, _rev: 1-643db5608d6cce93ad21003fe9e6e886] Saved as rev 1-643db5608d6cce93ad21003fe9e6e886
2015-02-16 18:16:37.923 [INFO ] tapshotpro[61493:607] TeamViewController:78 - Saved: [name: Habs, gameTypeFilter: 31, showInactivePlayers: 0, type: Team, _id: 14B53188-5AC0-451D-9104-4663108C3BDB, _rev: 1-643db5608d6cce93ad21003fe9e6e886]
18:16:37.924| CBLDatabase: Begin transaction (level 1)...
18:16:37.924| ViewVerbose:     kCblViewTeamsByName last indexed at #5
18:16:37.924| ViewVerbose: #6: map "14B53188-5AC0-451D-9104-4663108C3BDB" for view kCblViewTeamsByName...
18:16:37.924| ViewVerbose:     emit("Habs", {"_id":"14B53188-5AC0-451D-9104-4663108C3BDB","type":"Team","_local_seq":6,"name":"Habs","_rev":"1-643db5608d6cce93ad21003fe9e6e886","gameTypeFilter":31,"showInactivePlayers":false})
18:16:37.925| CBLDatabase: Commit transaction (level 1)
18:16:37.925| ViewVerbose: Query kCblViewTeamsByName: Will reduce row with key="Habs", value={"_id":"14B53188-5AC0-451D-9104-4663108C3BDB","type":"Team","_local_seq":6,"name":"Habs","_rev":"1-643db5608d6cce93ad21003fe9e6e886","gameTypeFilter":31,"showInactivePlayers":false}
18:16:37.925| ViewVerbose: Query kCblViewTeamsByName: Will reduce row with key="Leafs", value={"_id":"87CEECEB-C280-4389-80BD-CB65C270829D","type":"Team","_local_seq":5,"name":"Leafs","_rev":"1-649e533cfa9c4aa19a5fd996b2fd2674","gameTypeFilter":31,"showInactivePlayers":false}
18:16:37.925| ViewVerbose: Query kCblViewTeamsByName: Reduced to key=null, value=2
18:16:37.926| Query: CBLLiveQuery[kCblViewActivePlayersByNumber, start=["87CEECEB-C280-4389-80BD-CB65C270829D"], end=["87CEECEB-C280-4389-80BD-CB65C270829D","￿","￿"], mapOnly=YES]: Will update after 0 sec...
18:16:37.926| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: Will update after 0 sec...
18:16:37.926| Query: CBLLiveQuery[kCblViewActivePlayersByNumber, start=["87CEECEB-C280-4389-80BD-CB65C270829D"], end=["87CEECEB-C280-4389-80BD-CB65C270829D","￿","￿"], mapOnly=YES]: Async query tapshotpro/kCblViewActivePlayersByNumber...
18:16:37.926| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: Async query tapshotpro/kCblViewTeamsByName...
18:16:37.926‖ CBLDatabase: Begin transaction (level 1)...
18:16:37.926‖ ViewVerbose:     kCblViewActivePlayersByNumber last indexed at #5
18:16:37.927‖ ViewVerbose: #6: map "14B53188-5AC0-451D-9104-4663108C3BDB" for view kCblViewActivePlayersByNumber...
18:16:37.927‖ CBLDatabase: Commit transaction (level 1)
18:16:37.928| Query: CBLLiveQuery[kCblViewActivePlayersByNumber, start=["87CEECEB-C280-4389-80BD-CB65C270829D"], end=["87CEECEB-C280-4389-80BD-CB65C270829D","￿","￿"], mapOnly=YES]: ...async query finished (0 rows)
18:16:37.928‖ CBLDatabase: Begin transaction (level 1)...
18:16:37.928‖ CBLDatabase: Commit transaction (level 1)
18:16:37.928‖ ViewVerbose: Query kCblViewTeamsByName: Found row with key="Habs", value={"_id":"14B53188-5AC0-451D-9104-4663108C3BDB","type":"Team","_local_seq":6,"name":"Habs","_rev":"1-643db5608d6cce93ad21003fe9e6e886","gameTypeFilter":31,"showInactivePlayers":false}, id="14B53188-5AC0-451D-9104-4663108C3BDB"
18:16:37.929‖ ViewVerbose: Query kCblViewTeamsByName: Found row with key="Leafs", value={"_id":"87CEECEB-C280-4389-80BD-CB65C270829D","type":"Team","_local_seq":5,"name":"Leafs","_rev":"1-649e533cfa9c4aa19a5fd996b2fd2674","gameTypeFilter":31,"showInactivePlayers":false}, id="87CEECEB-C280-4389-80BD-CB65C270829D"
18:16:37.929| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: ...async query finished (2 rows)
18:16:37.929| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: ...Rows changed! (now 2)
2015-02-16 18:16:37.929 [INFO ] tapshotpro[61493:607] TeamsViewController:121 - Live query updated with 2 rows
2015-02-16 18:16:37.930 [INFO ] tapshotpro[61493:607] TeamsViewController:239 - Selecting team 87CEECEB-C280-4389-80BD-CB65C270829D at row 1 
```

**Comment by pasin** [comment]
@jpaas, thanks for getting us back. I reopened the issue.

From my first look at your both logs, the difference is that on your first lunch, the live query didn't get the database change notification (assuming that there is no more log messages after that).

**Comment by jpaas** [comment]
That is the end of the log for both launches. At least that is as much that is logged after adding the first model to the list.

**Comment by pasin** [comment]
@jpaas , I just have a chance to look at your comment and your log closely. 

You mentioned about view indexing on line 12:
```
18:14:55.537| Query: CBLLiveQuery[kCblViewTeamsByName, mapOnly=YES]: Will update after 0.434258 sec...
```
Normally the message `Will update after xxx sec` is logged when a CBLLiveQuery object gets a database change (kCBLDatabaseChangeNotification) notification. During your first launch, do you have another thread loading some document into the database while setting up the live query object?

**Comment by pasin** [comment]
Now I can reproduce the issue by adding a new document before setting a live query and the tableview datasource object.

**Comment by jpaas** [comment]
My app's startup sequence goes like this:

1. The initial view (MigrationViewController) is basically a splash screen where I set up the database.
2. When MigrationViewController appears, it sets up the main thread database instance and runs any database migrations.
3. MigrationViewController sets up the background database instance
4. MigrationViewController then replaces the rootViewController of the app (itself) with the TeamsViewController
5. The TeamsViewController is bound to a CBLUITableSource. In viewDidLoad, it creates the live query and assigns it to the table source.
6. Only once the TeamsViewController is loaded can any models in that live query be created or edited.

Having written this out now I'm wondering if setting up the background database instance is causing the problem. That is done using a serial dispatch queue. So its quite possible that step 3 is still processing while steps 4 and 5 are proceeding. I'll test this out by calling back to the main thread before proceeding with step 4.

**Comment by jpaas** [comment]
No the background database instance is not the issue. According to the log it completes before TeamsViewController.viewDidLoad is called. I even commented out the background db initialization and the TeamsViewController live query still doesn't update after adding the first model.

**Comment by pasin** [comment]
@jpaas Thanks a lot for all of the information you provided. I was able use your information to reproduce the issue. The issue was that if there is a document (or more) added prior to creating a live query, it could cause one of the flags (called willUpdate) inside the live query not get reset as a result the live query will stop updating itself when there are new documents put into the database. 

Anyway, the fix has been merged into the master branch. Could you check if it solves your issue?

**Comment by jpaas** [comment]
No problem at all. Thank you @pasin and @snej for always being so responsive to any reported issues. I'll try it out in the next couple of days and let you know.

Its true that I am creating a document before the live query is created. But its a document with a type that does not fall into the query, so I didn't really consider it. If this fixes it, I think it will fix at least a couple of bugs in my app.

**Comment by jpaas** [comment]
Confirmed that this now works in head(1.4). Thanks.
